flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3755,34 +3755,36 @@ class ArrayExprLowering {
 
       // Handle the negated conditions. See 10.2.3.2p4 as to why this control
       // structure is produced.
-      auto maskExprs = implicitSpace->getExprs();
-      const auto size = maskExprs.size() - 1;
-      for (std::remove_const_t<decltype(size)> i = 0; i < size; ++i)
-        if (maskExprs[i]) {
+      auto masks = implicitSpace->getMasks();
+      for (auto maskExprs : masks) {
+        const auto size = maskExprs.size() - 1;
+        for (std::remove_const_t<decltype(size)> i = 0; i < size; ++i)
+          if (maskExprs[i]) {
+            auto ifOp = builder.create<fir::IfOp>(
+                loc, mlir::TypeRange{innerArg.getType()},
+                fir::getBase(genCond(
+                    implicitSpace->getBindingWithShape(maskExprs[i]), iters)),
+                /*withElseRegion=*/true);
+            builder.create<fir::ResultOp>(loc, ifOp.getResult(0));
+            builder.setInsertionPointToStart(&ifOp.thenRegion().front());
+            builder.create<fir::ResultOp>(loc, innerArg);
+            builder.setInsertionPointToStart(&ifOp.elseRegion().front());
+          }
+
+        // The last condition is either non-negated or unconditionally negated.
+        if (maskExprs[size]) {
           auto ifOp = builder.create<fir::IfOp>(
               loc, mlir::TypeRange{innerArg.getType()},
               fir::getBase(genCond(
-                  implicitSpace->getBindingWithShape(maskExprs[i]), iters)),
+                  implicitSpace->getBindingWithShape(maskExprs[size]), iters)),
               /*withElseRegion=*/true);
           builder.create<fir::ResultOp>(loc, ifOp.getResult(0));
-          builder.setInsertionPointToStart(&ifOp.thenRegion().front());
-          builder.create<fir::ResultOp>(loc, innerArg);
           builder.setInsertionPointToStart(&ifOp.elseRegion().front());
+          builder.create<fir::ResultOp>(loc, innerArg);
+          builder.setInsertionPointToStart(&ifOp.thenRegion().front());
```

Comment:
Something looks off here. Landing in the `else` region is fully intended on the last condition, but this is placing the insertion point in the `then` region.

```Fortran
where (c1)
   ...
elsewhere (c2)
   ...
elsewhere
   <array-assignment> **
end where
```
Should generate the following control-flow structure
```
   implicit-loop-nest:
      if (cached-c1):
      else:
        if (cached-c2):
        else:
          <assign-element> **
```

Reply:
I think the diff is misleading and doesn't really show the change I did. Basically I just added the loop 
```
auto masks = implicitSpace->getMasks();
       for (auto maskExprs : masks) {
```
around the existing block. So instead of dealing with all the masks at once, we deal with each set of masks independently. The behavior in a single where (including elsewheres) construct is unchanged.

Reply:
Thanks. Makes sense. I'll take a closer look.